### PR TITLE
[FLINK-28586] Enable speculative execution for new sources

### PR DESCRIPTION
**Pull request by zhuzhurk**
## What is the purpose of the change

This PR enables new sources(FLIP-27) for speculative execution.
It mainly improved the SourceCoordinator and SourceCoordinatorContext to deal with attempt level split requests and source events.

## Verifying this change

  - *Added unit tests SourceCoordinatorConcurrentAttemptsTest*
  - *Added test cases in SourceCoordinatorTest*
  - *Covered by existing tests of SourceCoordinator and SourceCoordinatorContext*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java [hunk 1]
```diff
@@ -431,4 +489,125 @@ private <V> V callInCoordinatorThread(Callable<V> callable, String errorMessage)
...
+        getRegisteredAttempts(subtaskIndex)
+                .forEach(attempt -> assignSplitsToAttempt(subtaskIndex, attempt, splits));
+    }
+
+    private void assignSplitsToAttempt(int subtaskIndex, int attemptNumber, List<SplitT> splits) {
+        checkAttemptReaderReady(subtaskIndex, attemptNumber);
+
+        final AddSplitEvent<SplitT> addSplitEvent;
+        try {
+            addSplitEvent = new AddSplitEvent<>(splits, splitSerializer);
```
Comment (beyond1920) [inline flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java:507]: It's better not send this message if splits is empty or null.

**Review by beyond1920 — COMMENTED** [review]
Perfect, LTGM.
I only left a minor comment.

File: flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java (hunk 1 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java:511]: fixed.

**Review by beyond1920 — APPROVED** [review]

**Comment by zhuzhurk** [comment]
Also verified with TPC-DS benchmarks on real cluster. Speculative execution is working and there is no suspicious error.
Merging.

**Pull request closed without merging**